### PR TITLE
Remote data downloaders

### PR DESCRIPTION
**Pull request by Zmeos**

### CODE REVIEW

File: src/cobrabox/downloader.py [hunk 1]
```diff
@@ -0,0 +1,1021 @@
...
+import urllib.error
+import urllib.request
+import zipfile
+from collections.abc import Callable, Sequence
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from pathlib import Path
+from queue import Queue
+
+from tqdm import tqdm
```
Comment (jajcayn) [inline src/cobrabox/downloader.py:20]: rich!!!

Comment (Zmeos) [reply src/cobrabox/downloader.py:20]: Will add once we merge this, with a separate PR. I just want to get this merged.

**Comment by Zmeos** [comment]
Ok, ready for review.

**Review by jajcayn — APPROVED** [review]
## PR #88 Review: Remote Data Downloaders

### Summary

This PR introduces a **remote dataset downloading system** that enables CobraBox to download and load real EEG datasets from public repositories (PhysioNet, OpenNeuro, ETH Zurich, etc.). The implementation is well-architected, thoroughly tested, and production-ready.

---

### What This PR Does Well

#### 1. Clean Architecture
- **Separation of concerns**: `downloader.py` (download logic) <-> `datasets.py` (public API) <-> `dataset_loader.py` (format-specific loaders)
- **Lazy loading**: File index (`file_index.json`) loaded on-demand, not at import time
- **Extensible design**: Adding a new remote dataset requires only adding a `RemoteDatasetSpec` to `REMOTE_DATASETS`

#### 2. User Experience
- **Confirmation prompts**: Users see license, size, and file count before downloading (skippable with `accept=True`)
- **Subset support**: Download only specific subjects/files instead of entire datasets
- **Resume capability**: Interrupted downloads resume automatically via `.part` files + HTTP range requests
- **Cache validation**: Manifest-based integrity checking + ZIP validation for archives

#### 3. Error Handling
- **Specific error messages**: 401/403 shows `auth_hint`, timeout mentions 120s limit, ENOSPC suggests freeing disk space
- **KeyboardInterrupt handling**: Leaves `.part` files for resume, doesn't corrupt existing data
- **Defensive parsing**: All format loaders (CSV, NPY, MAT, EDF) have fallback behavior

#### 4. Testing
- **179 tests** covering downloader, datasets, and dataset_loader
- **Mocked network calls**: Tests don't require actual downloads
- **Edge cases**: Corrupt ZIPs, empty files, malformed JSON sidecars, HDF5 MAT files

#### 5. Documentation
- **Comprehensive docstrings**: Every public function has full parameter descriptions
- **Example usage**: `examples/remote_datasets_demo.py` shows all common patterns
- **Updated guides**: `docs/guide/datasets.md` includes remote dataset section

---

### Concerns and Suggestions

#### 1. File Index Size (CRITICAL)
The `file_index.json` is **23,757 lines** (~600KB) containing every single file URL for large datasets like CHB-MIT. This is:
- **Version control bloat**: Every URL change triggers a massive diff
- **Slow to load**: Parses the entire index even for small operations

**Recommendation**: Consider generating the file index dynamically from a smaller manifest, or host it remotely and cache locally.

#### 2. Circular Import Pattern (MINOR)
```python
def _swiss_eeg_short_spec() -> RemoteDatasetSpec:
    from .dataset_loader import _load_swiss_eeg_short  # avoid circular import
```
This is a code smell. The loaders import downloader types, and downloader specs import loader functions.

**Recommendation**: Consider a registry pattern where loaders register themselves, or move specs to a separate module.

#### 3. Hardcoded URLs (MEDIUM)
URLs are embedded in `file_index.json` and spec functions. If a repository moves (e.g., PhysioNet restructures), the code breaks.

**Recommendation**: Add a URL resolution layer or configuration override mechanism.

#### 4. MNE Dependency (MEDIUM)
EDF loading requires MNE-Python, which is a heavy dependency (~100MB). Currently it's pulled transitively by `mne-connectivity`, but this is fragile.

**Recommendation**: Add `mne` as an explicit optional dependency with a clear error message (already done, but should be in `pyproject.toml` `[project.optional-dependencies]`).

#### 5. License Compliance (MEDIUM)
Some datasets have restrictive licenses (e.g., "commercial and military use prohibited" for Swiss EEG). The prompt shows the license, but there's no audit trail.

**Recommendation**: Log license acceptances to `~/.cobrabox/acceptances.json` for compliance tracking.

---

### Code Quality Observations

| Aspect          | Rating       | Notes                                      |
| --------------- | ------------ | ------------------------------------------ |
| Type hints      | Excellent | Full coverage, no `Any` escapes              |
| Error messages  | Excellent | Actionable, specific, user-friendly        |
| Test coverage   | Excellent | 98% on new files, edge cases covered       |
| Documentation   | Excellent | Docstrings, examples, guides               |
| Maintainability | Good      | File index size is concerning              |
| Security        | Good      | No credential handling, but no audit trail |

---

### Breaking Changes

**YES** - The PR renames `cb.dataset()` to `cb.load_dataset()`. This is a **breaking change** for existing users.

**Recommendation**: Add a deprecation alias for one release cycle:
```python
def dataset(identifier: str) -> Dataset[SignalData]:
    """Deprecated: Use load_dataset() instead."""
    import warnings
    warnings.warn("cb.dataset() is deprecated. Use cb.load_dataset().", DeprecationWarning, stacklevel=2)
    return load_dataset(identifier)
```

---

### Verdict

**APPROVE with minor revisions**

This is a **high-quality, production-ready implementation** that significantly expands CobraBox's capabilities. The architecture is sound, testing is thorough, and the user experience is polished.

**Required before merge:**
1. Add deprecation alias for `cb.dataset()` -> `cb.load_dataset()`
2. Document the file index size concern and long-term plan

**Optional improvements:**
1. Add license acceptance logging
2. Make MNE an explicit optional dependency
3. Consider dynamic file index generation

---

### Stats

| Metric                           | Value   |
| -------------------------------- | ------- |
| Files changed                    | 32      |
| Additions                        | 29,046  |
| Deletions                        | 151     |
| New tests                        | 179     |
| Test coverage (new files)        | 85-98%  |
| Remote datasets                  | 7       |
| Total download size (all remote) | ~1.3 TB |

---

**Reviewer**: AI Agent (Sisyphus)  
**Review date**: 2026-04-07  
**Confidence**: High (thorough testing, clean architecture)

**Comment by jajcayn** [comment]
@Zmeos the review by LLM is ok-ish.. I get we need that HUGE JSON, so either try to implement what it suggests, or, at least, put in behind lfs
